### ISSUE TITLE
Using async validators with allErrors:true does not run in parallel

### ISSUE THREAD

**Issue by pselden** [ajv-validator/ajv#445]
I'm using { allErrors: true} with an async schema with async keywords.

(Side note: the reason I'm using allErrors: true is because I actually want to validate individual fields at runtime -- like I click off of a field on the UI and it runs the validation -- but I won't be able to get a specific field's errors back unless I run allErrors: true... correct? Is there a better way to do that?)

I noticed that the async keywords run sequentially and wait for the result of the previous one to finish, leading to a performance issue if you have multiple async validators that are waiting on some external resource.

**What version of Ajv are you using? Does the issue happen if you use the latest version?**
4.4..0 (issue still occurs in beta version)

**Ajv options object (see https://github.com/epoberezkin/ajv#options):**

```javascript
{ allErrors: true }
```

**JSON Schema (please make it as small as possible to reproduce the issue):**

```json
{
  "$async": true,
  "title": "User Schema",
  "type": "object",
  "properties": {
    "firstName": {
      "description": "First Name",
      "type": "string",
      "nameBlacklist": true
    },
    "lastName": {
      "description": "Last Name",
      "type": "string",
      "nameBlacklist": true
    }
  }
}
```

**Data (please make it as small as posssible to reproduce the issue):**

```json
{
  "firstName": "Paul",
  "lastName": "Paul"
}
```

**Your code (please use `options`, `schema` and `data` as variables):**

https://runkit.com/58d3cd89ae69180014e71b26/58d3cd89ae69180014e71b27

**What results did you expect?**
Expected time to take ~1000ms (running async keywords in parallel).

Actual time is ~2000ms (indicating sequential execution of keywords)

**Are you going to resolve the issue?**
If it is determined to be valid, then yes I can.

**Comment by epoberezkin** [comment]
@pselden thank you for highlighting the problem. I don't think there is any viable way to resolve it on Ajv level. All Ajv does is adding yields (or awaits) in front of async keyword calls so the code stays "synchronous" in structure. Doing anything more advanced is definitely out of scope - it would dramatically increase the complexity and given the relatively small share of users who use async validation it is not justified.

What you can do on your end is to define the keyword differently - rather than using it in each property to have it appear once on the top level with the list of properties. This way you can control how you do validation in your own code. It's the same syntax that `required` keyword uses:

```json
{
  "$async": true,
  "title": "User Schema",
  "type": "object",
  "nameBlacklist": ["firstName", "lastName"],
  "properties": {
    "firstName": {
      "description": "First Name",
      "type": "string",
    },
    "lastName": {
      "description": "Last Name",
      "type": "string"
    }
  }
}
```

**Comment by pselden** [comment]
Thanks for the speedy response! That example was just a trivial distillation of what we were going to attempt to do with it, so unfortunately changing the keywords won't work because we'll have multiple of them.

I'm working on a wrapper around a json schema validator that would be able to run on both front end and backend -- with the ability to specify server-side-only keywords for when we need to hit the DB. This requirement would essentially force _all_ validation to need to be async.

I think maybe what I could do to lessen the impact is to make it pass allErrors: false when running on the server, since "live" validation is only really useful on the client.

Edit: actually no, wouldn't be able to do allErrors: false because it will still _run_ the keywords synchronously.

**Comment by epoberezkin** [comment]
@pselden The only difference without allErrors: true (false is the default) is that it fails sooner, successful validation will take the same time, so yes it won't be faster.

The problem is essentially taking a "synchronous looking" code with yields (or awaits) and converting it into a code where all yields are resolved in parallel. This is not a trivial task. If you know how to do this transformation of the code - then you can supply Ajv with a different transpiler and it will do what you need.

Refactoring the schema can be not as simple as in the example above, but I think it should be possible. For example, you can have custom keywords that register necessary processing (and the registration itself is synchronous) and have another keyword that is only used on top level that executes all async operations in one go (or it can even be outside of schema) so that all your async operations are batched.

Another option is to use promises as a part of the data you validate and do merging of these promises (that's the approach I used in JSONScript).

So I am sure there are a couple of solutions that would give you parallel async operations with some simple changes in how you define and use custom keywords.

**Comment by mattfysh** [comment]
I'm running into this issue now, and am looking into replacing where `validateArray` uses a `gen.forRange`, and instead of:

```js
const len0 = data.length;
for (let i0 = 0; i0 < len0; i0++) { ... }
```

trying to generate:

```js
await Promise.all(data.map(async (_, i0) => { ... } ) 
```

Can you see any issues with this approach?

**Comment by mattfysh** [comment]
how I've ended up solving it is splitting into two phases with a couple of tricks to get it working:

Phase 1 is completely synchronous - I change the schema such that:

```
{
  type: 'object',
  keyword: true,
  properties: { ... }
}
```


becomes:

```
{
  anyOf: [
    { type: 'object', keyword: true },
    { type: 'object', properties: { ... } },
  ]
}
```

And then inside the validate function for `keyword`, I return `true` if the function has resulted in some asynchronous work, while updating the value to be the promise. If the workload was not asynchronous, i return `false` so that the next entry in the `anyOf` array is run

Phase 2 is the asynchronous piece, where all the promises created in Phase 1 are awaited. The new schema used is (notice `allOf` this time instead of `anyOf`)

```
{
  $async: true,
  allOf: [
    { type: 'object', keyword: true },
    { type: 'object', properties: { ... } },
  ]
}
```

In the validate function for keyword, I check if the incoming value is a promise, and if it is then I `await` it, then use the resolved value to update the parent property

For the purpose of triggering multiple asynchronous workloads so they may be batched in the next tick of the event loop, this appears to be working quite well, despite the mental gymnastics required to build it.
